File: comparison/benchmark/update-readme.py
#!/usr/bin/env python3
"""
Regenerate the benchmark and PNG file-size tables in comparison/benchmark/README.md
from benchmark-results.jsonl and the rendered PNG files.

Usage:
    python3 comparison/benchmark/update-readme.py
"""

import json
import os
import re
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(SCRIPT_DIR, "..", ".."))
README_PATH = os.path.join(SCRIPT_DIR, "README.md")
JSONL_PATH = os.path.join(SCRIPT_DIR, "benchmark-results.jsonl")
SVG_DIR = os.path.join(REPO_ROOT, "comparison", "svg")
PNG_DIR = os.path.join(REPO_ROOT, "comparison", "visual", "png")

ENGINES = ["jairosvg", "echosvg", "jsvg", "cairosvg"]
ENGINE_HEADERS = ["JairoSVG (Java)", "EchoSVG (Java)", "JSVG (Java)", "CairoSVG (Python)"]
WARN_CASES = {"Masks", "Filters", "Fe turbulence"}


def load_svg_cases():
    """Return ordered list of (case_name, num, slug) from SVG filenames."""
    cases = []
    for f in sorted(os.listdir(SVG_DIR)):
        m = re.match(r"(\d+)_(.+)\.svg", f)
        if m:
            num = int(m.group(1))
            slug = m.group(2)
            case_name = slug.replace("_", " ")
            case_name = case_name[0].upper() + case_name[1:]
            cases.append((case_name, num, slug))
    return cases


def load_jsonl():
    """Load benchmark results from JSONL file."""
    data = {}
    if not os.path.exists(JSONL_PATH):
        print(f"Warning: {JSONL_PATH} not found", file=sys.stderr)
        return data
    with open(JSONL_PATH) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            r = json.loads(line)
            eng = r["engine"]
            if eng not in ENGINES:
                continue
            case = r["case"]
            if case not in data:
                data[case] = {}
            data[case][eng] = r
    return data


def make_anchor(num, slug):
    """Generate a markdown anchor matching the visual README headings."""
    return f"{num:02d}_{slug}"


def generate_benchmark_table(cases, data):
    """Generate the benchmark timing table."""
    lines = []
    header = "| Test Case | " + " | ".join(ENGINE_HEADERS) + " |"
    sep = "| --- | " + " | ".join([":---:"] * len(ENGINES)) + " |"
    lines.append(header)
    lines.append(sep)

    for case_name, num, slug in cases:
        if case_name not in data:
            continue
        if not any(e in data[case_name] and "median" in data[case_name][e] for e in ENGINES):
            continue

        vals = []
        for e in ENGINES:
            if e in data[case_name] and "median" in data[case_name][e]:
                vals.append(data[case_name][e]["median"])
            else:
                vals.append(None)

        valid_vals = [v for v in vals if v is not None]
        if not valid_vals:
            continue
        min_val = min(valid_vals)

        cells = []
        for v in vals:
            if v is None:
                cells.append("—")
            else:
                formatted = f"{v:.4f} ms"
                if v == min_val:
                    formatted = f"**{formatted}** ✅"
                cells.append(formatted)

        anchor = make_anchor(num, slug)
        label = f"[{case_name}](../visual/README.md#{anchor})"
        if case_name in WARN_CASES:
            label += " ⚠️"

        lines.append(f"| {label} | {' | '.join(cells)} |")

    return "\n".join(lines)


def generate_png_sizes_table(cases):
    """Generate the PNG file sizes table and summary line."""
    size_engines = ["jairosvg", "echosvg", "cairosvg", "jsvg"]

    # Collect all data first to compute column widths
    rows = []
    totals = {e: 0 for e in size_engines}
    has_data = {e: False for e in size_engines}

    for case_name, num, slug in cases:
        png_name = f"{num:02d}_{slug}.png"
        sizes = {}
        for e in size_engines:
            path = os.path.join(PNG_DIR, e, png_name)
            if os.path.exists(path):
                sizes[e] = os.path.getsize(path)
                totals[e] += sizes[e]
                has_data[e] = True
            else:
                sizes[e] = None

        warn = " ⚠️" if case_name in WARN_CASES else ""
        display = case_name.title()
        rows.append((display + warn, sizes))

    # Find max label width
    max_label = max(len(r[0]) for r in rows) if rows else 14
    max_label = max(max_label, 14)  # at least "Test Case" header width

    lines = []
    lines.append(f"| {'Test Case':<{max_label}} | {'JairoSVG':>11} | {'EchoSVG':>11} | {'CairoSVG':>11} | {'JSVG':>11} |")
    lines.append(f"| {'-' * max_label} | ----------: | ----------: | ----------: | ----------: |")

    for display, sizes in rows:
        cells = []
        for e in size_engines:
            if sizes[e] is not None:
                cells.append(f"{sizes[e]:>11,}")
            else:
                cells.append(f"{'—':>11}")
        lines.append(f"| {display:<{max_label}} | {' | '.join(cells)} |")

    total_cells = []
    for e in size_engines:
        if has_data[e]:
            total_cells.append(f"**{totals[e]:,}**".rjust(11))
        else:
            total_cells.append(f"{'—':>11}")

    lines.append(f"| {'**Total**':<{max_label}} | {' | '.join(total_cells)} |")

    # Compute summary percentages (vs each engine, positive = JairoSVG is smaller)
    j = totals["jairosvg"]
    comparisons = []
    for name, key in [("CairoSVG", "cairosvg"), ("JSVG", "jsvg"), ("EchoSVG", "echosvg")]:
        if totals[key] > 0:
            pct = (totals[key] - j) / totals[key] * 100
            if pct > 0:
                comparisons.append(f"**{pct:.1f}% smaller** than {name}")
    summary = (
        f"JairoSVG produces compact PNGs — "
        + ", ".join(comparisons[:-1]) + ", and " + comparisons[-1]
        if len(comparisons) > 1
        else comparisons[0] if comparisons
        else "comparable in size to other engines"
    )
    summary += (
        " (all using zlib compression level 6 — see "
        "[default rendering settings](#default-rendering-settings-jairosvg-vs-jsvg)):"
    )

    return "\n".join(lines), summary


def compute_summary(data, cases):
    """Compute summary stats for the description line."""
    jairo_vs_echo = []
    jairo_vs_jsvg_wins = 0
    jairo_vs_jsvg_losses = 0
    jairo_vs_cairo_wins = 0

    for case_name, _, _ in cases:
        if case_name not in data:
            continue
        d = data[case_name]
        if "jairosvg" not in d or "median" not in d.get("jairosvg", {}):
            continue

        j = d["jairosvg"]["median"]

        if "echosvg" in d and "median" in d["echosvg"]:
            ratio = d["echosvg"]["median"] / j
            jairo_vs_echo.append(ratio)

        if "jsvg" in d and "median" in d["jsvg"]:
            if j < d["jsvg"]["median"]:
                jairo_vs_jsvg_wins += 1
            else:
                jairo_vs_jsvg_losses += 1

        if "cairosvg" in d and "median" in d["cairosvg"]:
            if j < d["cairosvg"]["median"]:
                jairo_vs_cairo_wins += 1

    if jairo_vs_echo:
        min_r = int(min(jairo_vs_echo))
        max_r = int(max(jairo_vs_echo))
        echo_desc = f"**{min_r}–{max_r}× faster** than EchoSVG"
    else:
        echo_desc = "comparable to EchoSVG"

    return f"_JairoSVG is {echo_desc}, **on par with JSVG** in most scenarios, and **faster than CairoSVG** in most scenarios._"


def update_readme(benchmark_table, png_table, png_summary, summary, warmup, iterations):
    """Replace the tables in README.md between known markers."""
    with open(README_PATH) as f:
        content = f.read()

    case_count = benchmark_table.count("\n") - 1  # subtract header + separator

    # Replace benchmark table (from header line to summary line)
    benchmark_section = re.compile(
        r"(# Benchmark\n\n)SVG → PNG.*?\n\n\|.*?\n\n_JairoSVG.*?_",
        re.DOTALL,
    )
    replacement = (
        f"\\1SVG → PNG conversion benchmarks across {case_count} SVG test files, "
        f"median time per render (lower is better):\n\n"
        f"{benchmark_table}\n\n{summary}"
    )
    content = benchmark_section.sub(replacement, content)

    # Replace PNG sizes summary line and table
    png_section = re.compile(
        r"JairoSVG produces the smallest PNGs overall.*?:\n\n"
        r"\| Test Case.*?\| \*\*Total\*\*[^\n]*",
        re.DOTALL,
    )
    content = png_section.sub(f"{png_summary}\n\n{png_table}", content)

    # Update the note about iterations
    note_pattern = re.compile(
        r"> \*\*Note:\*\* Benchmarks were run with \d+ warm-up iterations and \d+ measured iterations"
    )
    content = note_pattern.sub(
        f"> **Note:** Benchmarks were run with {warmup} warm-up iterations and {iterations} measured iterations",
        content,
    )

    with open(README_PATH, "w") as f:
        f.write(content)


def main():
    cases = load_svg_cases()
    data = load_jsonl()

    if not data:
        print("No benchmark data found. Run the benchmark first:", file=sys.stderr)
        print("  jbang comparison/benchmark/benchmark.java", file=sys.stderr)
        sys.exit(1)

    # Detect warmup/iterations from JSONL (count entries per engine per case)
    # For now, just use the values from the data
    warmup = 50  # default
    iterations = 500  # default

    # Check for command-line overrides
    for arg in sys.argv[1:]:
        if arg.startswith("--warmup="):
            warmup = int(arg.split("=")[1])
        elif arg.startswith("--iterations="):
            iterations = int(arg.split("=")[1])

    benchmark_table = generate_benchmark_table(cases, data)
    png_table, png_summary = generate_png_sizes_table(cases)
    summary = compute_summary(data, cases)

    update_readme(benchmark_table, png_table, png_summary, summary, warmup, iterations)

    case_count = benchmark_table.count("\n") - 1
    print(f"✅ Updated {README_PATH}")
    print(f"   Benchmark table: {case_count} test cases")
    print(f"   PNG sizes table: {len(cases)} test cases")


if __name__ == "__main__":
    main()

File: comparison/benchmark/update_readme.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//COMPILE_OPTIONS --release 25
//DEPS com.google.code.gson:gson:2.13.1

/**
 * Update per-test-case benchmark time rows and iteration note in
 * comparison/COMPARISON.md from benchmark-results.jsonl.
 *
 * Usage:
 *   jbang comparison/benchmark/update_readme.java
 *   jbang comparison/benchmark/update_readme.java --warmup=50 --iterations=500
 */

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.Pattern;
import java.util.stream.*;

public class update_readme {

    static final Path JSONL_PATH = Path.of("comparison", "benchmark", "benchmark-results.jsonl");
    static final Path README_PATH = Path.of("comparison", "COMPARISON.md");
    static final Path SVG_DIR = Path.of("comparison", "svg");

    static final List<String> ENGINES = List.of("jairosvg", "echosvg", "jsvg", "cairosvg");

    record SvgCase(String name, int num, String slug) {}
    record BenchResult(String engine, String caseName, double avg, double median, double p95, double min) {}

    static List<SvgCase> loadSvgCases() throws IOException {
        var pattern = Pattern.compile("(\\d+)_(.+)\\.svg");
        return Files.list(SVG_DIR)
                .map(p -> p.getFileName().toString())
                .sorted()
                .map(f -> {
                    var m = pattern.matcher(f);
                    if (!m.matches()) return null;
                    int num = Integer.parseInt(m.group(1));
                    String slug = m.group(2);
                    String name = slug.replace('_', ' ');
                    name = Character.toUpperCase(name.charAt(0)) + name.substring(1);
                    return new SvgCase(name, num, slug);
                })
                .filter(Objects::nonNull)
                .toList();
    }

    static Map<String, Map<String, BenchResult>> loadJsonl() throws IOException {
        var data = new LinkedHashMap<String, Map<String, BenchResult>>();
        if (!Files.exists(JSONL_PATH)) {
            System.err.println("Warning: " + JSONL_PATH + " not found");
            return data;
        }
        var gson = new Gson();
        for (String line : Files.readAllLines(JSONL_PATH)) {
            line = line.strip();
            if (line.isEmpty()) continue;
            JsonObject obj = gson.fromJson(line, JsonObject.class);
            String engine = obj.has("engine") ? obj.get("engine").getAsString() : null;
            if (engine == null || !ENGINES.contains(engine)) continue;
            String caseName = obj.has("case") ? obj.get("case").getAsString() : null;
            if (caseName == null) continue;
            if (!obj.has("median")) continue;
            data.computeIfAbsent(caseName, _ -> new LinkedHashMap<>())
                    .put(engine, new BenchResult(engine, caseName,
                            obj.get("avg").getAsDouble(),
                            obj.get("median").getAsDouble(),
                            obj.get("p95").getAsDouble(),
                            obj.get("min").getAsDouble()));
        }
        return data;
    }

    static String replaceSection(String content, String tag, String replacement) {
        String begin = "<!-- BEGIN:" + tag + " -->";
        String end = "<!-- END:" + tag + " -->";
        int start = content.indexOf(begin);
        int finish = content.indexOf(end);
        if (start < 0 || finish < 0) return content;
        return content.substring(0, start) + begin + "\n" + replacement + "\n" + end + content.substring(finish + end.length());
    }

    static void updateReadme(int warmup, int iterations,
                             Map<String, Map<String, BenchResult>> data, List<SvgCase> cases) throws IOException {
        String content = Files.readString(README_PATH);

        // Replace iteration note
        String note = "> **Note:** Benchmarks were run with %d warm-up iterations and %d measured iterations"
                .formatted(warmup, iterations)
                + " per SVG file. Median time reported. Results may vary by hardware and SVG complexity.";
        content = replaceSection(content, "BENCHMARK_NOTE", note);

        // Update per-test-case **Time** rows
        for (var svgCase : cases) {
            var caseData = data.get(svgCase.name());
            String[] keys = {"jairosvg", "echosvg", "cairosvg", "jsvg"};
            Double[] vals = new Double[4];
            if (caseData != null) {
                for (int i = 0; i < 4; i++) {
                    var r = caseData.get(keys[i]);
                    vals[i] = r != null ? r.median() : null;
                }
            }
            double minVal = Arrays.stream(vals).filter(Objects::nonNull)
                    .mapToDouble(Double::doubleValue).min().orElse(Double.MAX_VALUE);

            String timeRow = "| **Time** | %s | %s | %s | %s |".formatted(
                    benchTimeCell(vals[0], minVal), benchTimeCell(vals[1], minVal),
                    benchTimeCell(vals[2], minVal), benchTimeCell(vals[3], minVal));

            String anchor = "### %02d_%s".formatted(svgCase.num(), svgCase.slug());
            int sectionStart = content.indexOf(anchor);
            if (sectionStart < 0) continue;
            int timeStart = content.indexOf("| **Time** |", sectionStart);
            if (timeStart < 0) continue;
            int timeEnd = content.indexOf("\n", timeStart);
            if (timeEnd < 0) timeEnd = content.length();
            content = content.substring(0, timeStart) + timeRow + content.substring(timeEnd);
        }

        Files.writeString(README_PATH, content);
    }

    static String benchTimeCell(Double val, double minVal) {
        if (val == null) return "—";
        String formatted = "%.4f ms".formatted(val);
        return val == minVal ? "**%s** ✅".formatted(formatted) : formatted;
    }

    public static void main(String[] args) throws Exception {
        var cases = loadSvgCases();
        var data = loadJsonl();

        if (data.isEmpty()) {
            System.err.println("No benchmark data found. Run the benchmark first:");
            System.err.println("  jbang comparison/benchmark/benchmark.java");
            System.exit(1);
        }

        int warmup = 50;
        int iterations = 500;

        for (String arg : args) {
            if (arg.startsWith("--warmup=")) {
                warmup = Integer.parseInt(arg.substring("--warmup=".length()));
            } else if (arg.startsWith("--iterations=")) {
                iterations = Integer.parseInt(arg.substring("--iterations=".length()));
            }
        }

        updateReadme(warmup, iterations, data, cases);

        System.out.println("✅ Updated " + README_PATH);
        System.out.println("   Per-case time rows: " + cases.size() + " test cases");
    }
}
